[Chore/YB-541]: 데이터베이스 설정(MySQL8)

Ywoosang opened the pull request:
[//]: # (PR 제목 예시 : [Feature/YB-1]: 소셜 로그인 기능 구현)

## 📄 Work Description
<strong>Jira Ticket : `YB-541`</strong>
<!--
해당 PR에서 어떤 작업을 진행했는지 알려주세요 (코드, 이미지 첨부를 통해 설명해주시면 이해하기 더 쉬울것 같아요!)

[예시]
1. page, size 최솟값 설정 및 Unit Test 추가
    기존의 코드에는 page와 size의 범위 설정을 해주지 않아서 page가 1보다 작은 경우 에러가 발생했습니다. 이를 방지하기 위해 page와 size 둘다 1 이상의 정수만 전달 받을 수 있도록 설정했습니다.
    (코드 및 이미지 첨부)
-->
## 작업 내용
#### 1. mysql8.x 버전으로 데이터베이스 변경
논의한 대로, 유지보수가 용이하도록 익숙한 mysql8.0 으로 데이터베이스 변경했습니다.

#### 2. RepositoryTest 테스트에서도 TestContainers 를 이용하도록 설정
기존에는 통합테스트에서만 Testcontainers 를 사용하고, RepositoryTest 에서는 h2 데이터베이스를 이용하고 있었습니다.
이를 Testcontainers 를 이용하는 방식으로 통일했습니다.

#### 3. 테스트간 격리 깨지는 문제 해결
RepositoryTest 클래스에서 `@SQL` 로 cleanup.sql 을 이용해 데이터를 지워주었는데, 주문 도메인의 RepositoryTest 의 `@SQL` 과 충돌이 발생해 데이터가 지워지지 않는 문제점이 있었습니다.

테스트에서 `@DataJpaTest` 애노테이션을 사용하고 있기 때문에 내부에 `@Transactional` 이 있어, 기본적으로 테스트 종료시 데이터가 롤백됩니다.

여기에 LawQuery 로 AutoIncrement 값만 추가적으로 초기화시켜cleanup.sql 없이 정상적으로 데이터 초기화가 이루어지도록 변경했습니다.

#### 4. 테스트코드에서 TestEntityManager 를 이용하도록 통일
리포지토리 테스트마다 TestEntityManager 를 사용하는 테스트가있고, EntityManager 를 주입받아 사용하는 테스트도 있어 메소드를 호출할때 혼동이 발생했습니다.

예를들어 TestEntityManager 의 persist 는 Entity 를 반환하는데, EntityManager 는 반환값이 void 입니다.

TestEntityManager 를 사용하는 방향이 코드 길이가 줄어들고 초기화 로직이 깔끔해지기 떄문에 TestEntityManager 로 통일했습니다.

#### 5. Flyway 마이그레이션

기존 migration history 를 초기화하고 V0__init.sql (schema.sql 역할)  부터 시작하도록 재구성했습니다.  

#### 6. Participant 엔티티와 Member 엔티티의 중복되는 role 문제 해결
동일한 이름의 role 때문에 혼동이 발생해, 팀별 사용자 권한은 team_member_role 로 변경하고 Flyway 에 반영했습니다.


#### 7. 리드미에 ERD 추가
변경된 Column 반영해 erd 사진 추가했습니다.

Ywoosang committed:
chore(YB-545): mysql 로 데이터베이스 변경
- flyway 및 테스트용 sql 스크립트 문법 변경

Ywoosang committed:
chore: mysql 로 설정 변경

Ywoosang committed:
chore(YB-546): 통합테스트 jwtUtil 을 testFixtures 로 이동

Ywoosang committed:
feat(YB-547): 통합테스트시 컨테이너 하나만 사용하도록 testcontainers 설정 변경

Ywoosang committed:
feat(YB-548): Repository 테스트에도 testcontainers 를 사용하도록 설정

Ywoosang committed:
merge: develop 변경사항 반영

Ywoosang committed:
chore(YB-555): 테스트 의존성에 롬복 추가

Ywoosang committed:
chore(YB-556): Repository 테스트 리팩토링

Ywoosang committed:
chore: 인프라 폴더명 간소화
- infrastructure -> infra

Ywoosang committed:
refactor(YB-557): 중복되는 role 이름 수정

Ywoosang committed:
remove: 불필요한 cleanup.sql 제거
- member 엔티티 주석 제거 포함

Ywoosang committed:
chore: domain 모듈 테스트에서 sql 파일 위치를 resources/sql 로 이동

Ywoosang committed:
feat(YB-558): 개발시 swagger 에서 팀별 사용자 권한을 볼 수 있도록 추가

Ywoosang committed:
docs: readme 에 erd 추가

jorippppong commented on yellobook-domain/src/test/java/com/yellobook/support/RepositoryTest.java:
```diff
@@ -0,0 +1,77 @@
+        registry.add("spring.datasource.password", mysql::getPassword);
+    }
+
+    protected void resetAutoIncrement() {
```
기존의 cleanup.sql 에서 auto increment를 설정해주었는데, 해당 부분을 여기서 진행하면 어떤 부분의 차이가 있나요?

Ywoosang replied:
sql - TRUNCATE, AUTO_INCREMENT 초기화
 
sql 스크립트를 제거하고, 데이터를 지우는 작업은 `@DataJpaTest` 에 있는 `@Transactional` 을 이용하도록 변경했어요resetAutoIncrement 메서드로AUTO_INCREMENT 만 초기화하도록 설정한 부분입니다.

jorippppong commented:
`@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)`

를 사용하면 테스트 간 충돌이 없어지고 문제 상황을 해결할 수 있는데, 매번 테스트가 끝나고 애플리케이션 컨텍스트가 "더티(dirtied)" 상태로 표시되서 이후 테스트에서 애플리케이션 컨텍스트가 다시 로드되어 초기 상태로 돌아가는게 시간이  엄청 많이 걸리네요..! 이 방법은 비추...
<img width="734" alt="스크린샷 2024-10-28 오후 3 17 07" src="https://github.com/user-attachments/assets/16b03191-0697-4359-a76d-2ba2acfb215b">

Ywoosang commented:
@jorippppong 
맞아요 저 방법 쓰면 컨텍스트 매번 새로 로드해야해서 격리는 보장되는데 엄청 오래걸리더라구요 ㅜ

Ywoosang committed:
chore: 테스트시 트랜잭션 롤백과 @Sql 실행 로그 확인 가능하도록 설정

Ywoosang committed:
chore: jdbcTemplate을 이용해 매 테스트 메서드 종료 이후 데이터 초기화하도록 변경
- 롤백 후에도 여전히 데이터 남아있는 문제 해결

Ywoosang committed:
test(YB-559): Repository 테스트에서 @DataJpaTest 롤백 기능 활용하도록 변경

Ywoosang committed:
chore: RepositoryTest 에서 jdbcTemplate 로그 출력하도록 설정

Ywoosang committed:
test: 테스트에서 반환타입 일치시킴
- SonarCloud 정적분석 결과 참고 (Change the assertion arguments to not compare dissimilar types)
